### PR TITLE
slave: send ack to controller on opportunity steering request

### LINKED ISSUE #986: [TASK] Handle Client Steering Opportunity
## Description

The Client Steering message can be a Mandate or an Opportunity. Currently only the Mandate is handled. We need to handle the Opportunity as well.

## Exit criteria

- [ ] test_flows includes a test with Opportunity (cfr. certification test 4.8.3).
- [ ] Agent sends an ACK when receiving Client Steering Opportunity.
- [ ] Agent sends a Steering Completed when receiving Client Steering Opportunity.
- [ ] Agent stores the Opportunity Window for the BSS as a deadline timestamp.

### PR DESCRIPTION
**Pull request by RanRegev**
When receiving a Steering Opportunity request, reply to the
controller with an ACK message before sending the Steering
Completed message.

### CODE REVIEW

**Comment by morantr** [comment]
LGTM, If that all the changes mark this PR as "ready for review" and I will approve.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (ghost) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: Please remove TODOs when the work is TODONE :slightly_smiling_face:

**Review by ghost — CHANGES_REQUESTED** [review]
Please add or change a flow test to test this feature

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (vitalybu) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: This flow is already tested as part of the certification tests.
The goal of this PR is to respond with the appropriate messages to the controller (to pass the certification) and not to actually implement the steering opportunity logic (which is not yet defined for prplMesh).

File: .gitignore
Comment (vitalybu) [inline .gitignore]: That's OK, but should be on a separate commit

Comment (RanRegev) [reply .gitignore]: If it is fine, I would prefer to skip the separation for this time only.

Comment (rmelotte) [reply .gitignore]: Actually this change is not ok.
See Arnout's excellent comment here: https://github.com/prplfoundation/prplMesh/pull/308#discussion_r332587456

This is something that has nothing to do with prplMesh, but is only related to the editor you use.
You can configure it locally either for a specific repository or globally.

Comment (vitalybu) [reply .gitignore]: Makes sense.
@RanRegev, please remove this commit.

Comment (RanRegev) [reply .gitignore]: > Actually this change is not ok.
> See Arnout's excellent comment here: [#308 (comment)](https://github.com/prplfoundation/prplMesh/pull/308#discussion_r332587456)
> 
> This is something that has nothing to do with prplMesh, but is only related to the editor you use.
> You can configure it locally either for a specific repository or globally.
Done according to the recommendation in #308 (comment).

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp:4382]: Excellent comment! :trophy:

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This comment, on the other hand, is redundant. When you create a message of type ACK_MESSAGE, it's fairly obvious that it's an ACK message. No need to add a comment for that.

**Review by arnout — APPROVED** [review]
Looks good to me. Is there any reason why this is still draft?

If it is just because you want to do more testing - no reason to keep it draft for that. It won't get merged until you set the Ready for merge label.

That said, you really should add a second commit that updates test_flows.py to test this flow.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (regevran) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: During the years I found it very useful to read the comments _separately_ from the code.
So in this else clause, if one reads the first comment, and then the two successive comments - there is a story, the _intent_ is clear and well described. One can tackle the intent without reading a single line of code. The next thing to do is to _verify_ the code follows the comments.
In this way we separated the logic from implementation and helped building correct and maintainable code.

Comment (RanRegev) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: I found it very useful to read the comments separately from the code.
So in this else clause, if one reads the first comment, and then the two successive comments - there is a story, the intent is clear and well described. One can tackle the intent without reading a single line of code. The next thing to do is to verify the code follows the comments.
In this way we separated the logic from implementation and helped building correct and maintainable code.

**Comment by RanRegev** [comment]
> Please add or change a flow test to test this feature

**Review by tomereli — CHANGES_REQUESTED** [review]
Merge commits are forbidden - please refer to the CONTRIBUTING.md and do a git interactive rebase.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Still - it is redundant, and this is how most if not all the code is written, so please remove the comment.

**Comment by tomereli** [comment]
Regarding the test - please add a separate test for this flow.

**Comment by RanRegev** [comment]
> Please add or change a flow test to test this feature

Already implemented:
https://github.com/prplfoundation/prplMesh/blob/master/tests/test_flows.py#L559-L572

**Review by vitalybu — APPROVED** [review]

**Review by tomereli — APPROVED** [review]
LGTM but did it pass the actual test? If so, please add a comment to the PR description with the link to the gitlab pipeline which passed the test, this should be merged only after this - so removing the ready for merge tag till this is done.

**Comment by rmelotte** [comment]
Here is a passing certification test on this branch: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/487143085

**Review by ghost — APPROVED** [review]

**Pull request merged**
